fms/training/plugins.py

```diff
@@ -0,0 +1,228 @@
+from abc import abstractmethod
+from datetime import datetime
+import os
+from pathlib import Path
+from typing import Dict, List
+import torch
+from torch import distributed as dist
+from torch import nn
+from fms import utils
+from fms.utils import print0, generation
+
+
+class TrainerPlugin:
+    """
+    A TrainerPlugin runs once every epoch, and possibly every `steps` steps.
+    It is passed relevant objects that can be used for checkpointing, logging,
+    or validation.
+    """
+
+    def __init__(self, steps=None):
+        self.steps = steps
+
+    def run(self, step):
+        """
+        Whether or not to run this plugin on the current step.
+        """
+        # if step is None, we're at an epoch end not an intermediate step.
+        # By default we always run for epoch ends.
+        if step is None:
+            return True
+        # If self.steps is None, we're only recording epoch ends and this isn't one.
+        if self.steps is None:
+            return False
+        # record every `step` steps
+        if (step + 1) % self.steps == 0:
+            return True
+        return False
+
+    @abstractmethod
+    def step(
+        self,
+        model: nn.Module,
+        optimizer,
+        epoch: int,
+        metrics: Dict = {},
+        step: int = None,
+    ):
+        pass
+
+
+class InferenceValidator(TrainerPlugin):
+    """
+    A training plugin to print the results of running inference on a given prompt.
+    """
+
+    def __init__(
+        self, prompt_tokens: List[str], tokenizer, device, steps=None, eos_token=None
+    ):
+        super().__init__(steps)
+        self.tokenizer = tokenizer
+        input_ids = tokenizer.convert_tokens_to_ids(prompt_tokens)
+        self.input_ids = torch.tensor(input_ids, dtype=torch.long, device=device)
+        self.eos_token_id = (
+            None
+            if eos_token is None
+            else self.tokenizer.convert_tokens_to_ids([eos_token])[0]
+        )
+
+    def step(
+        self,
+        model: nn.Module,
+        optimizer,
+        epoch: int,
+        metrics: Dict = {},
+        step: int = None,
+    ):
+        if not self.run(step):
```

Comment:
should this somehow be getting called implicitly? I would expect that this function would be running outside somewhere and if returning true, would call this step function

Reply:
see other response re MetricReporter